### ISSUE TITLE
facingMode problem

### ISSUE THREAD

**Issue by Mario8419** [kybarg/react-qr-scanner#39]
Hello and thank you for giving this repo a new boost.
It's quite working good except the
facingMode property isn't working through various devices. It's always the front camera which is active despite the prop is set to rear. Any suggestion?

```
<QrReader
     facingMode="rear"
     delay={500}
     onError={handleError}
     onScan={handleScan}
     style={{ width: '100%' }}
/>
```

**Comment by PusangMaySaging** [comment]
@Mario8419  I have the same problem. Mine is throwing this error whenever i pass the facingMode prop
![image](https://user-images.githubusercontent.com/54404521/108615456-396ae300-743f-11eb-8b69-e8a6516d936d.png)
`          <div className="scanner-wrapper">
                    <span ref={scannerMessageRef}>Ready to Scan</span>
                    <QrReader
                        onScan={handleHandleScan}
                        facingMode="rear"
                        onError={handleError}
                        className="scanner"
                        
                    ></QrReader>
                    <div className=""></div>
                </div>`

**Comment by alatnet** [comment]
Tacking on this issue with that deviceId has the same thing.

**Comment by senda-dev** [comment]
This repo is for desktop, if you want mobile support, look for: https://github.com/JodusNodus/react-qr-reader

**Comment by alatnet** [comment]
Doesnt work on desktop.  I've tried it with desktop.

**Comment by Mario8419** [comment]
demo here: https://kybarg.github.io/react-qr-scanner/ is working, also mobile

**Comment by kevin-ping** [comment]
to make the code show no error, you need to put 'rear into {}, like below

<QrReader
     facingMode={"rear"}
     delay={500}
....

HOWEVER, this is not working either.

senda-dev is correct, once you change to JodusNodus repo, everyting works.

**Comment by shendel** [comment]
npm repo contains the wrong version, different from the code here

**Comment by shendel** [comment]
`npm i https://github.com/kybarg/react-qr-scanner` fix problem templonary, but this is not good solution, need update package in npm

**Comment by ernest-okot** [comment]
after painfully reading through the compiled code, I found that you can pass media constraints to the component

```js
<QrReader
  delay={1000}
  style={{
    width: "100%",
    height: "100%"
  }}
  constraints={
    isDesktop
      ? undefined
      : {
          video: {
            facingMode: { exact: facingMode }
          }
        }
  }
  facingMode={facingMode}
  onError={handleError}
  onScan={handleScan}
/>
```

**Comment by ernest-okot** [comment]
@shendel the git version doesn't seem to work

**Comment by shendel** [comment]
@ernest-okot  really? ) 
see for yourself and make sure
`npm i react-qr-scanner@1.0.0-alpha.7`
https://www.npmjs.com/package/react-qr-scanner

![image](https://user-images.githubusercontent.com/1880211/113456708-aa29f580-9416-11eb-9021-b103112183e7.png)
Really?
Face, Rare?
Front, Back?
![image](https://user-images.githubusercontent.com/1880211/113456797-ea897380-9416-11eb-9031-7242286b67a6.png)

May be you see `getDeviceId` function there????

For me full-link repo from this git-hub repo work fine, with front-back camera, with select between front-back

main problem - npm version is not up to date

**Comment by shendel** [comment]
@ernest-okot https://github.com/kybarg/react-qr-scanner/blob/master/lib/index.js#L394
Any questions?
Where `facingMode`, `legacyMode`??? Where `chooseDeviceId` func????

**Comment by shendel** [comment]
@ernest-okot , yes - github build dont scan ((( ... may be we need wait owner of repo for answers....

**Comment by shendel** [comment]
@kybarg , @fjogeleit , can you are help us and fix this?

**Comment by shendel** [comment]
@ernest-okot i make fix https://github.com/kybarg/react-qr-scanner/pull/41
you can check on my fork https://github.com/shendel/react-qr-scanner

**Comment by kybarg** [comment]
@shendel
1. this projecxt is mainly focused for desktp browsers, for mobile devices I'd sugest https://github.com/JodusNodus/react-qr-reader
2. **react-qr-scanner@1.0.0-alpha.7** doesn't have legacy support yet
3. Facing mode and device id schould be passed through [cosntraints](https://developer.mozilla.org/en-US/docs/Web/API/Media_Streams_API/Constraints)

**Comment by shendel** [comment]
@kybarg , ничего что я по русски, у вас в профиле указан Киев. На английском сложно высказываться. Да я понимаю, но в мобильном браузере тоже работает, плюс этот репо хоть как-то обновляется в отличии от того, что вы указали, что, по крайней мере для меня является приоритетом. И да - были опечатки. Не распознавало код (я поправил и сделал пул)

**Comment by deepak3387** [comment]
facingMode property is not working for me as well. 
npm repo is also confusing and how to use property is not clear.
I would suggest to kindly add some example with all the available properties so that its easy to use.
Also I would love to see an example containing option for selecting available camera(s).

**Comment by bansalshashank** [comment]
hello its not working for me as well is there anyone for whom its working if not then does anybody found out any solution for the same

**Comment by deepak3387** [comment]
@bansalshashank 
I was able to solve this with below code:

```
const [width, setWidth] = useState(window.innerWidth);

function handleWindowSizeChange() {
    setWidth(window.innerWidth);
}

useEffect(() => {
    window.addEventListener('resize', handleWindowSizeChange);
    return () => {
        window.removeEventListener('resize', handleWindowSizeChange);
    }
}, []);

let isDesktop = (width > 768);

<QrReader
  delay={500}
  style={styles.previewStyle}
  onError={handleError}
  onScan={handleScan}
  constraints={
      isDesktop
      ? undefined
      : {
          video: {
              facingMode: { exact: `environment` }
          }
        }
  }
/>
```
Hope this helps!

**Comment by HeinCorbet** [comment]
I found the problem with facingMode not wiorking for me is because the camera label is language specific.  In Dutch the camera says 'voorzijde' (=front) and 'achterzijde (=back). Adding a new property to supply additional filter values could be a solution. Providing a select camera option to the end user seems to be the way to go, that is what I did. I found however that the qr-scanner does not switch camera when changing a state value. It only seems to rerender when changing the prop 'delay', 'facingMode' and 'legacymode' How about adding a property to set the camera deviceId when known.  I now solved this by putting the deviceId in the facingMode property to trigger the rerendering (switching) to the other camera. This triggers the 'chooseDeviceId' function where I can return the selected deviceId.
